In [1]:
import pandas as pd
melb_data = pd.read_csv('data/melb_data_ps.csv', sep=',')

# Базовые операции со столбцами DataFrame #

### Создание копии таблицы ###

In [2]:
melb_df = melb_data.copy()
display(melb_df)

index         Suburb  ... Propertycount           Coordinates
0          0     Abbotsford  ...          4019    -37.7996, 144.9984
1          1     Abbotsford  ...          4019    -37.8079, 144.9934
2          2     Abbotsford  ...          4019    -37.8093, 144.9944
3          3     Abbotsford  ...          4019    -37.7969, 144.9969
4          4     Abbotsford  ...          4019    -37.8072, 144.9941
...      ...            ...  ...           ...                   ...
13575  13575  Wheelers Hill  ...          7392  -37.90562, 145.16761
13576  13576   Williamstown  ...          6380  -37.85927, 144.87904
13577  13577   Williamstown  ...          6380  -37.85274, 144.88738
13578  13578   Williamstown  ...          6380  -37.85908, 144.89299
13579  13579     Yarraville  ...          6543  -37.81188, 144.88449

[13580 rows x 23 columns]

In [3]:
melb_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13580 entries, 0 to 13579
Data columns (total 23 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   index          13580 non-null  int64  
 1   Suburb         13580 non-null  object 
 2   Address        13580 non-null  object 
 3   Rooms          13580 non-null  int64  
 4   Type           13580 non-null  object 
 5   Price          13580 non-null  float64
 6   Method         13580 non-null  object 
 7   SellerG        13580 non-null  object 
 8   Date           13580 non-null  object 
 9   Distance       13580 non-null  float64
 10  Postcode       13580 non-null  int64  
 11  Bedroom        13580 non-null  int64  
 12  Bathroom       13580 non-null  int64  
 13  Car            13580 non-null  int64  
 14  Landsize       13580 non-null  float64
 15  BuildingArea   13580 non-null  float64
 16  YearBuilt      13580 non-null  int64  
 17  CouncilArea    12211 non-null  object 
 18  Lattit

### Удаление столбцов ###

In [4]:
melb_df.drop(['index', 'Coordinates'], axis='columns', inplace=True)
melb_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13580 entries, 0 to 13579
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Suburb         13580 non-null  object 
 1   Address        13580 non-null  object 
 2   Rooms          13580 non-null  int64  
 3   Type           13580 non-null  object 
 4   Price          13580 non-null  float64
 5   Method         13580 non-null  object 
 6   SellerG        13580 non-null  object 
 7   Date           13580 non-null  object 
 8   Distance       13580 non-null  float64
 9   Postcode       13580 non-null  int64  
 10  Bedroom        13580 non-null  int64  
 11  Bathroom       13580 non-null  int64  
 12  Car            13580 non-null  int64  
 13  Landsize       13580 non-null  float64
 14  BuildingArea   13580 non-null  float64
 15  YearBuilt      13580 non-null  int64  
 16  CouncilArea    12211 non-null  object 
 17  Lattitude      13580 non-null  float64
 18  Longti

### Математические операции со столбцами ###

In [5]:
total_rooms = melb_df['Rooms'] + melb_df['Bedroom'] + melb_df['Bathroom']
display(total_rooms)

0         5
1         5
2         8
3         8
4         8
         ..
13575    10
13576     8
13577     8
13578     9
13579     9
Length: 13580, dtype: int64

In [6]:
melb_df['MeanRoomSquare'] = melb_df['BuildingArea'] / total_rooms
display(melb_df['MeanRoomSquare'])

0        25.200000
1        15.800000
2        18.750000
3        15.750000
4        17.750000
           ...    
13575    12.600000
13576    16.625000
13577    15.750000
13578    17.444444
13579    12.444444
Name: MeanRoomSquare, Length: 13580, dtype: float64

In [7]:
diff_area = melb_df['BuildingArea'] - melb_df['Landsize']
sum_area = melb_df['BuildingArea'] + melb_df['Landsize']
melb_df['AreaRatio'] = diff_area / sum_area
display(melb_df['AreaRatio'])

0       -0.231707
1       -0.327660
2        0.056338
3        0.145455
4        0.083969
           ...   
13575   -0.676093
13576   -0.429185
13577   -0.551601
13578   -0.693060
13579   -0.527426
Name: AreaRatio, Length: 13580, dtype: float64

# Работа с датами в DataFrame #

### Формат datetime ###

In [8]:
display(melb_df['Date'])

0         3/12/2016
1         4/02/2016
2         4/03/2017
3         4/03/2017
4         4/06/2016
            ...    
13575    26/08/2017
13576    26/08/2017
13577    26/08/2017
13578    26/08/2017
13579    26/08/2017
Name: Date, Length: 13580, dtype: object

In [9]:
melb_df['Date'] = pd.to_datetime(melb_df['Date'], dayfirst=True)
display(melb_df['Date'])
melb_df.info()

0       2016-12-03
1       2016-02-04
2       2017-03-04
3       2017-03-04
4       2016-06-04
           ...    
13575   2017-08-26
13576   2017-08-26
13577   2017-08-26
13578   2017-08-26
13579   2017-08-26
Name: Date, Length: 13580, dtype: datetime64[ns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13580 entries, 0 to 13579
Data columns (total 23 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   Suburb          13580 non-null  object        
 1   Address         13580 non-null  object        
 2   Rooms           13580 non-null  int64         
 3   Type            13580 non-null  object        
 4   Price           13580 non-null  float64       
 5   Method          13580 non-null  object        
 6   SellerG         13580 non-null  object        
 7   Date            13580 non-null  datetime64[ns]
 8   Distance        13580 non-null  float64       
 9   Postcode        13580 non-null  int64         
 10  Bedroom         13580 non-null  int64         
 11  Bathroom        13580 non-null  int64         
 12  Car             13580 non-null  int64         
 13  Landsize        13580 non-null  float64       
 14  BuildingArea    13580 non-null  float64       
 15  Ye

### Выделение атрибутов datetime ###

In [10]:
years_sold = melb_df['Date'].dt.year
print(years_sold)
print('Min year sold', years_sold.min())
print('Max year sold', years_sold.max())
print('Mode year sold', years_sold.mode()[0])

0        2016
1        2016
2        2017
3        2017
4        2016
         ... 
13575    2017
13576    2017
13577    2017
13578    2017
13579    2017
Name: Date, Length: 13580, dtype: int64
Min year sold 2016
Max year sold 2017
Mode year sold 2017


In [11]:
melb_df['MonthSale'] = melb_df['Date'].dt.month
melb_df.info()
melb_df['MonthSale'].value_counts(normalize=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13580 entries, 0 to 13579
Data columns (total 24 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   Suburb          13580 non-null  object        
 1   Address         13580 non-null  object        
 2   Rooms           13580 non-null  int64         
 3   Type            13580 non-null  object        
 4   Price           13580 non-null  float64       
 5   Method          13580 non-null  object        
 6   SellerG         13580 non-null  object        
 7   Date            13580 non-null  datetime64[ns]
 8   Distance        13580 non-null  float64       
 9   Postcode        13580 non-null  int64         
 10  Bedroom         13580 non-null  int64         
 11  Bathroom        13580 non-null  int64         
 12  Car             13580 non-null  int64         
 13  Landsize        13580 non-null  float64       
 14  BuildingArea    13580 non-null  float64       
 15  Ye

5     0.149411
7     0.145950
9     0.135862
6     0.134757
8     0.114138
11    0.082032
4     0.069882
3     0.049926
12    0.044698
10    0.040574
2     0.032622
1     0.000147
Name: MonthSale, dtype: float64

### Работа с интервалами ###

In [12]:
delta_days = melb_df['Date'] - pd.to_datetime('2016-01-01')
display(delta_days)
display(delta_days.dt.days)

0       337 days
1        34 days
2       428 days
3       428 days
4       155 days
          ...   
13575   603 days
13576   603 days
13577   603 days
13578   603 days
13579   603 days
Name: Date, Length: 13580, dtype: timedelta64[ns]

0        337
1         34
2        428
3        428
4        155
        ... 
13575    603
13576    603
13577    603
13578    603
13579    603
Name: Date, Length: 13580, dtype: int64

In [13]:
melb_df['AgeBuilding'] = melb_df['Date'].dt.year - melb_df['YearBuilt']
display(melb_df.info())
display(melb_df['AgeBuilding'])

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13580 entries, 0 to 13579
Data columns (total 25 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   Suburb          13580 non-null  object        
 1   Address         13580 non-null  object        
 2   Rooms           13580 non-null  int64         
 3   Type            13580 non-null  object        
 4   Price           13580 non-null  float64       
 5   Method          13580 non-null  object        
 6   SellerG         13580 non-null  object        
 7   Date            13580 non-null  datetime64[ns]
 8   Distance        13580 non-null  float64       
 9   Postcode        13580 non-null  int64         
 10  Bedroom         13580 non-null  int64         
 11  Bathroom        13580 non-null  int64         
 12  Car             13580 non-null  int64         
 13  Landsize        13580 non-null  float64       
 14  BuildingArea    13580 non-null  float64       
 15  Ye

None

0         46
1        116
2        117
3         47
4          2
        ... 
13575     36
13576     22
13577     20
13578     97
13579     97
Name: AgeBuilding, Length: 13580, dtype: int64

In [14]:
melb_df = melb_df.drop('YearBuilt', axis=1)
melb_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13580 entries, 0 to 13579
Data columns (total 24 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   Suburb          13580 non-null  object        
 1   Address         13580 non-null  object        
 2   Rooms           13580 non-null  int64         
 3   Type            13580 non-null  object        
 4   Price           13580 non-null  float64       
 5   Method          13580 non-null  object        
 6   SellerG         13580 non-null  object        
 7   Date            13580 non-null  datetime64[ns]
 8   Distance        13580 non-null  float64       
 9   Postcode        13580 non-null  int64         
 10  Bedroom         13580 non-null  int64         
 11  Bathroom        13580 non-null  int64         
 12  Car             13580 non-null  int64         
 13  Landsize        13580 non-null  float64       
 14  BuildingArea    13580 non-null  float64       
 15  Co

----

### Задание 3.3 ###

Создать в таблице `melb_df` признак `WeekdaySale` (день недели). Найти, сколько объектов недвижимости было продано в выходные (суббота и воскресенье), результат занести в переменную `weekend_count`.

In [15]:
melb_df['WeekdaySale'] = melb_df['Date'].dt.weekday
display(melb_df.info())
display(melb_df['WeekdaySale'])
holiday_mask = melb_df['WeekdaySale'] > 4
weekend_count = holiday_mask.sum()
print('Weekend Count = ', weekend_count)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13580 entries, 0 to 13579
Data columns (total 25 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   Suburb          13580 non-null  object        
 1   Address         13580 non-null  object        
 2   Rooms           13580 non-null  int64         
 3   Type            13580 non-null  object        
 4   Price           13580 non-null  float64       
 5   Method          13580 non-null  object        
 6   SellerG         13580 non-null  object        
 7   Date            13580 non-null  datetime64[ns]
 8   Distance        13580 non-null  float64       
 9   Postcode        13580 non-null  int64         
 10  Bedroom         13580 non-null  int64         
 11  Bathroom        13580 non-null  int64         
 12  Car             13580 non-null  int64         
 13  Landsize        13580 non-null  float64       
 14  BuildingArea    13580 non-null  float64       
 15  Co

None

0        5
1        3
2        5
3        5
4        5
        ..
13575    5
13576    5
13577    5
13578    5
13579    5
Name: WeekdaySale, Length: 13580, dtype: int64

Weekend Count =  12822


----

# Вводные данные для выполнения заданий 3.4-3.5 #

Представлены данные (в формате csv) об отчётах очевидцев НЛО в США за период с 1930 по 2020 год.

В данных есть следующие признаки:
```text
"City" — город, где был замечен НЛО;
"Colors Reported" — цвет объекта;
"Shape Reported" — форма объекта;
"State" — обозначение штата;
"Time" — время, когда был замечен НЛО (данные отсортированы от старых наблюдений к новым). 
```
Прочитать данные, сделать преобразование времени к формату `datetime` и выполнить задания ниже.

In [16]:
ufo_df = pd.read_csv('data/ufo.csv', sep=',')
display(ufo_df)
display(ufo_df.info())
print('Shape = ', ufo_df.shape)

City Colors Reported Shape Reported State              Time
0                    Ithaca             NaN       TRIANGLE    NY    6/1/1930 22:00
1               Willingboro             NaN          OTHER    NJ   6/30/1930 20:00
2                   Holyoke             NaN           OVAL    CO   2/15/1931 14:00
3                   Abilene             NaN           DISK    KS    6/1/1931 13:00
4      New York Worlds Fair             NaN          LIGHT    NY   4/18/1933 19:00
...                     ...             ...            ...   ...               ...
18236            Grant Park             NaN       TRIANGLE    IL  12/31/2000 23:00
18237           Spirit Lake             NaN           DISK    IA  12/31/2000 23:00
18238           Eagle River             NaN            NaN    WI  12/31/2000 23:45
18239           Eagle River             RED          LIGHT    WI  12/31/2000 23:45
18240                  Ybor             NaN           OVAL    FL  12/31/2000 23:59

[18241 rows x 5 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18241 entries, 0 to 18240
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   City             18216 non-null  object
 1   Colors Reported  2882 non-null   object
 2   Shape Reported   15597 non-null  object
 3   State            18241 non-null  object
 4   Time             18241 non-null  object
dtypes: object(5)
memory usage: 712.7+ KB


None

Shape =  (18241, 5)


In [17]:
ufo_df['Time'] = pd.to_datetime(ufo_df['Time'])
display(ufo_df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18241 entries, 0 to 18240
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   City             18216 non-null  object        
 1   Colors Reported  2882 non-null   object        
 2   Shape Reported   15597 non-null  object        
 3   State            18241 non-null  object        
 4   Time             18241 non-null  datetime64[ns]
dtypes: datetime64[ns](1), object(4)
memory usage: 712.7+ KB


None

----

###  Задание 3.4 ###

В каком году отмечается наибольшее количество случаев наблюдения НЛО в США?

In [18]:
year_only = ufo_df['Time'].dt.year
display(year_only.mode())

0    1999
Name: Time, dtype: int64

----

###  Задание 3.5 ###

Найти средний интервал времени (в днях) между двумя последовательными случаями наблюдения НЛО в штате Невада (NV).

    Чтобы выделить дату из столбца `Time`, можно воспользоваться атрибутом `datetime` `date`.
    Чтобы вычислить разницу между двумя соседними датами в столбце, примените к нему метод `diff()`.
    Чтобы перевести интервал времени в дни, воспользуйтесь атрибутом `timedelta` `days`.


In [19]:
nv_df = ufo_df[ufo_df['State'] == 'NV']
display(nv_df)
dates = nv_df['Time'].dt.date.diff()
print('Mean interval for Nevada =', round(dates.dt.days.mean()), 'days')

City Colors Reported Shape Reported State                Time
76      Las Vegas             NaN           DISK    NV 1947-07-15 10:00:00
172    Nellis AFB             NaN           DISK    NV 1952-02-17 18:00:00
565        Fallon             NaN           OVAL    NV 1959-09-15 00:00:00
566     Goldfield             NaN          LIGHT    NV 1959-09-15 01:00:00
613           NaN             NaN           DISK    NV 1960-07-01 12:00:00
...           ...             ...            ...   ...                 ...
17447    Laughlin             NaN      FORMATION    NV 2000-09-16 22:00:00
17567   Las Vegas             NaN         SPHERE    NV 2000-09-30 22:25:00
17617   Las Vegas      RED YELLOW          OTHER    NV 2000-10-06 20:25:00
17890        Reno             NaN       TRIANGLE    NV 2000-11-07 02:15:00
18104   Las Vegas             NaN       TRIANGLE    NV 2000-12-10 17:30:00

[284 rows x 5 columns]

Mean interval for Nevada = 69 days


# Создание и преобразование столбцов с помощью функций #

In [20]:
melb_df['Address'].nunique()

13378

In [21]:
print(melb_df['Address'].loc[177])
print(melb_df['Address'].loc[1812])
print(melb_df['Address'].loc[9001])

2/119 Railway St N
9/400 Dandenong Rd
172 Danks St


In [22]:
def get_street_type(address):
    exclude_list = ['N', 'S', 'E', 'W']
    address_list = address.split(' ')
    street_type = address_list[-1]
    if street_type in exclude_list:
        street_type = address_list[-2]
    return street_type

street_types = melb_df['Address'].apply(get_street_type)
display(street_types)
print(street_types.nunique(), 'уникальных значений')

0        St
1        St
2        St
3        La
4        St
         ..
13575    Cr
13576    Dr
13577    St
13578    St
13579    St
Name: Address, Length: 13580, dtype: object

56 уникальных значений


In [23]:
display(street_types.value_counts())

St           8012
Rd           2825
Ct            612
Dr            447
Av            321
Gr            311
Pde           211
Pl            169
Cr            152
Cl            100
La             67
Bvd            53
Tce            47
Wy             40
Avenue         40
Cct            25
Hwy            24
Parade         15
Boulevard      13
Sq             11
Crescent        9
Cir             7
Strand          7
Esplanade       6
Grove           5
Gdns            4
Grn             4
Fairway         4
Mews            4
Crossway        3
Righi           3
Victoria        2
Ridge           2
Crofts          2
Esp             2
Glade           1
Gra             1
Ave             1
Woodland        1
Outlook         1
Hts             1
Highway         1
Athol           1
Summit          1
Grand           1
Res             1
Nook            1
Eyrie           1
Dell            1
East            1
Loop            1
Grange          1
Terrace         1
Cove            1
Qy              1
Corso     

In [24]:
popular_stypes = street_types.value_counts().nlargest(10).index
print(popular_stypes)

Index(['St', 'Rd', 'Ct', 'Dr', 'Av', 'Gr', 'Pde', 'Pl', 'Cr', 'Cl'], dtype='object')


In [25]:
melb_df['StreetType'] = street_types.apply(lambda x: x if x in popular_stypes else 'other')
display(melb_df['StreetType'])
print('Unique = ', melb_df['StreetType'].nunique())

0           St
1           St
2           St
3        other
4           St
         ...  
13575       Cr
13576       Dr
13577       St
13578       St
13579       St
Name: StreetType, Length: 13580, dtype: object

Unique =  11


In [26]:
melb_df = melb_df.drop('Address', axis=1)
melb_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13580 entries, 0 to 13579
Data columns (total 25 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   Suburb          13580 non-null  object        
 1   Rooms           13580 non-null  int64         
 2   Type            13580 non-null  object        
 3   Price           13580 non-null  float64       
 4   Method          13580 non-null  object        
 5   SellerG         13580 non-null  object        
 6   Date            13580 non-null  datetime64[ns]
 7   Distance        13580 non-null  float64       
 8   Postcode        13580 non-null  int64         
 9   Bedroom         13580 non-null  int64         
 10  Bathroom        13580 non-null  int64         
 11  Car             13580 non-null  int64         
 12  Landsize        13580 non-null  float64       
 13  BuildingArea    13580 non-null  float64       
 14  CouncilArea     12211 non-null  object        
 15  La

----

###  Задание 4.2 ###

Написать функцию `get_weekend(weekday)`, которая принимает на вход элемент столбца `WeekdaySale` и возвращает 1, если день является выходным, и 0 — в противном случае, и создать столбец `Weekend` в таблице `melb_df` с помощью неё.

Применить эту функцию к столбцу и вычислить среднюю цену объекта недвижимости, проданного в выходные дни. Результат округлить до целых.

In [27]:
def get_weekend(weekday):
    if weekday > 4:
        return 1
    return 0

melb_df['Weekend'] = melb_df['WeekdaySale'].apply(get_weekend)

In [28]:
weekend_sales = melb_df[melb_df['Weekend'] == 1]
round(weekend_sales['Price'].mean())

1081199

----

###  Задание 4.3 ###

Преобразовать столбец `SellerG` с наименованиями риелторских компаний в таблице `melb_df` следующим образом: оставить в столбце только 49 самых популярных компаний, а остальные обозначить как `other`.

Найти, во сколько раз минимальная цена объектов недвижимости, проданных компанией 'Nelson', больше минимальной цены объектов, проданных компаниями, обозначенными как `other`. Ответ округлить до десятых.

In [29]:
print(melb_df['SellerG'].nunique())
best_sellers = melb_df['SellerG'].value_counts().nlargest(49).index
melb_df['SellerG'] = melb_df['SellerG'].apply(lambda x: x if x in best_sellers else 'other')
display(melb_df['SellerG'])
print(melb_df['SellerG'].nunique())

268


0          Biggin
1          Biggin
2          Biggin
3          Biggin
4          Nelson
           ...   
13575       Barry
13576    Williams
13577       Raine
13578     Sweeney
13579     Village
Name: SellerG, Length: 13580, dtype: object

50


In [30]:
nelson_min = melb_df[melb_df['SellerG'] == 'Nelson'].Price.min()
other_min = melb_df[melb_df['SellerG'] == 'other'].Price.min()
round(nelson_min / other_min, 1)

1.3

# Тип данных Category #

In [31]:
unique_list = []
for col in melb_df.columns:
    item = (col, melb_df[col].nunique(), melb_df[col].dtypes)
    unique_list.append(item)
unique_counts = pd.DataFrame(
    unique_list,
    columns=['Column_Name', 'Num_Unique', 'Type']
).sort_values(by='Num_Unique', ignore_index=True)
display(unique_counts)

Column_Name  Num_Unique            Type
0          Weekend           2           int64
1             Type           3          object
2           Method           5          object
3      WeekdaySale           5           int64
4       Regionname           8          object
5         Bathroom           9           int64
6            Rooms           9           int64
7              Car          11           int64
8       StreetType          11          object
9          Bedroom          12           int64
10       MonthSale          12           int64
11     CouncilArea          33          object
12         SellerG          50          object
13            Date          58  datetime64[ns]
14     AgeBuilding         152           int64
15        Postcode         198           int64
16        Distance         202         float64
17   Propertycount         311           int64
18          Suburb         314          object
19    BuildingArea         602         float64
20  MeanRoomSquare        1277         float64
21        Landsize        1448         float64
22           Price        2204         float64
23       Lattitude        6503         float64
24       AreaRatio        6521         float64
25      Longtitude        7063         float64

In [32]:
display(melb_df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13580 entries, 0 to 13579
Data columns (total 26 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   Suburb          13580 non-null  object        
 1   Rooms           13580 non-null  int64         
 2   Type            13580 non-null  object        
 3   Price           13580 non-null  float64       
 4   Method          13580 non-null  object        
 5   SellerG         13580 non-null  object        
 6   Date            13580 non-null  datetime64[ns]
 7   Distance        13580 non-null  float64       
 8   Postcode        13580 non-null  int64         
 9   Bedroom         13580 non-null  int64         
 10  Bathroom        13580 non-null  int64         
 11  Car             13580 non-null  int64         
 12  Landsize        13580 non-null  float64       
 13  BuildingArea    13580 non-null  float64       
 14  CouncilArea     12211 non-null  object        
 15  La

None

In [33]:
cols_to_exclude = ['Date', 'Rooms', 'Bedroom', 'Bathroom', 'Car']
max_unique_count = 150
for col in melb_df.columns:
    if melb_df[col].nunique() < max_unique_count and col not in cols_to_exclude:
        melb_df[col] = melb_df[col].astype('category')
display(melb_df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13580 entries, 0 to 13579
Data columns (total 26 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   Suburb          13580 non-null  object        
 1   Rooms           13580 non-null  int64         
 2   Type            13580 non-null  category      
 3   Price           13580 non-null  float64       
 4   Method          13580 non-null  category      
 5   SellerG         13580 non-null  category      
 6   Date            13580 non-null  datetime64[ns]
 7   Distance        13580 non-null  float64       
 8   Postcode        13580 non-null  int64         
 9   Bedroom         13580 non-null  int64         
 10  Bathroom        13580 non-null  int64         
 11  Car             13580 non-null  int64         
 12  Landsize        13580 non-null  float64       
 13  BuildingArea    13580 non-null  float64       
 14  CouncilArea     12211 non-null  category      
 15  La

None

In [34]:
print(melb_df['Regionname'].cat.categories)

Index(['Eastern Metropolitan', 'Eastern Victoria', 'Northern Metropolitan',
       'Northern Victoria', 'South-Eastern Metropolitan',
       'Southern Metropolitan', 'Western Metropolitan', 'Western Victoria'],
      dtype='object')


In [35]:
display(melb_df['Regionname'].cat.codes)

0        2
1        2
2        2
3        2
4        2
        ..
13575    4
13576    6
13577    6
13578    6
13579    6
Length: 13580, dtype: int8

In [36]:
melb_df['Type'] = melb_df['Type'].cat.rename_categories({
    'u': 'unit',
    't': 'townhouse',
    'h': 'house'
})
display(melb_df['Type'])

0        house
1        house
2        house
3        house
4        house
         ...  
13575    house
13576    house
13577    house
13578    house
13579    house
Name: Type, Length: 13580, dtype: category
Categories (3, object): ['house', 'townhouse', 'unit']

In [37]:
melb_df['Type'] = melb_df['Type'].cat.add_categories('flat')
new_houses_types = pd.Series(['unit', 'house', 'flat', 'flat', 'house'])
new_houses_types = new_houses_types.astype(melb_df['Type'].dtype)
display(new_houses_types)

0     unit
1    house
2     flat
3     flat
4    house
dtype: category
Categories (4, object): ['house', 'townhouse', 'unit', 'flat']

----

###  Задание 5.2 ###

При преобразовании столбцов таблицы о недвижимости к типу `category` остался без внимания столбец `Suburb` (пригород). Предлагается исправить это.

1. С помощью метода `info()` узнать, сколько памяти занимает таблица `melb_df`.
2. Преобразовать признак **Suburb** следующим образом: оставить в столбце только 119 наиболее популярных пригородов, остальные заменить на **'other'**.
3. Привести данные в столбце `Suburb` к категориальному типу.

В качестве ответа записать разницу между объёмом занимаемой памяти до преобразования (который мы получили ранее в модуле) и после него в Мб. Ответ округлить до десятых.

***Ответ - 0.1***

In [38]:
melb_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13580 entries, 0 to 13579
Data columns (total 26 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   Suburb          13580 non-null  object        
 1   Rooms           13580 non-null  int64         
 2   Type            13580 non-null  category      
 3   Price           13580 non-null  float64       
 4   Method          13580 non-null  category      
 5   SellerG         13580 non-null  category      
 6   Date            13580 non-null  datetime64[ns]
 7   Distance        13580 non-null  float64       
 8   Postcode        13580 non-null  int64         
 9   Bedroom         13580 non-null  int64         
 10  Bathroom        13580 non-null  int64         
 11  Car             13580 non-null  int64         
 12  Landsize        13580 non-null  float64       
 13  BuildingArea    13580 non-null  float64       
 14  CouncilArea     12211 non-null  category      
 15  La

In [39]:
popular_subs = melb_df['Suburb'].value_counts().nlargest(119).index
melb_df['Suburb'] = melb_df['Suburb'].apply(lambda x: x if x in popular_subs else 'other')
melb_df['Suburb'] = melb_df['Suburb'].astype('category')
display(melb_df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13580 entries, 0 to 13579
Data columns (total 26 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   Suburb          13580 non-null  category      
 1   Rooms           13580 non-null  int64         
 2   Type            13580 non-null  category      
 3   Price           13580 non-null  float64       
 4   Method          13580 non-null  category      
 5   SellerG         13580 non-null  category      
 6   Date            13580 non-null  datetime64[ns]
 7   Distance        13580 non-null  float64       
 8   Postcode        13580 non-null  int64         
 9   Bedroom         13580 non-null  int64         
 10  Bathroom        13580 non-null  int64         
 11  Car             13580 non-null  int64         
 12  Landsize        13580 non-null  float64       
 13  BuildingArea    13580 non-null  float64       
 14  CouncilArea     12211 non-null  category      
 15  La

None

# Задания по датасету велоспорта #

Датасет представляет собой таблицу с информацией о 300 тысячах поездок за первые пять дней сентября 2018 года и включает в себя следующую информацию:
- **`starttime`** — время начала поездки (дата, время);
- **`stoptime`** — время окончания поездки (дата, время);
- **`start station id`** — идентификатор стартовой стоянки;
- **`start station name`** — название стартовой стоянки;
- **`start station latitude`**, **`start station longitude`** — географическая широта и долгота стартовой стоянки;
- **`end station id`** — идентификатор конечной стоянки;
- **`end station name`** — название конечной стоянки;
- **`end station latitude`**, **`end station longitude`** — географическая широта и долгота конечной стоянки;
- **`bikeid`** — идентификатор велосипеда;
- **`usertype`** — тип пользователя (Customer — клиент с подпиской на 24 часа или на три дня, Subscriber — подписчик с годовой арендой велосипеда);
- **`birth year`** — год рождения клиента;
- **`gender`** — пол клиента (0 — неизвестный, 1 — мужчина, 2 — женщина).

In [42]:
bike_data = pd.read_csv('data/citibike-tripdata.csv', sep=',')
bike_df = bike_data.copy()
display(bike_df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300000 entries, 0 to 299999
Data columns (total 14 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   starttime                300000 non-null  object 
 1   stoptime                 300000 non-null  object 
 2   start station id         299831 non-null  float64
 3   start station name       299831 non-null  object 
 4   start station latitude   300000 non-null  float64
 5   start station longitude  300000 non-null  float64
 6   end station id           299831 non-null  float64
 7   end station name         299831 non-null  object 
 8   end station latitude     300000 non-null  float64
 9   end station longitude    300000 non-null  float64
 10  bikeid                   300000 non-null  int64  
 11  usertype                 300000 non-null  object 
 12  birth year               300000 non-null  int64  
 13  gender                   300000 non-null  int64  
dtypes: f

None

----

###  Задание 6.1 ###

Сколько пропусков в столбце `start station id`?

***Ответ - 169***

----

###  Задание 6.2 ###

Какой тип данных имеют столбцы `starttime` и `stoptime`?

***Ответ - object***

----

###  Задание 6.3 ###

Найти идентификатор самой популярной стартовой стоянки. Записать идентификатор в виде целого числа.

***Ответ - 281***

In [47]:
print(bike_df['start station id'].mode())
display(bike_df['start station id'].value_counts())

0    281.0
Name: start station id, dtype: float64


281.0     1928
2006.0    1909
519.0     1873
514.0     1845
426.0     1835
          ... 
3432.0      15
3599.0      11
2005.0       9
3704.0       7
3583.0       6
Name: start station id, Length: 759, dtype: int64

----

### Задание 6.4 ###

Велосипед с каким идентификатором является самым популярным?

***Ответ - 33887***

In [48]:
print(bike_df['bikeid'].mode())
display(bike_df['bikeid'].value_counts())

0    33887
Name: bikeid, dtype: int64


33887    107
34358    104
34159    103
34258    103
33916    102
        ... 
18587      1
17324      1
29593      1
33157      1
25794      1
Name: bikeid, Length: 10004, dtype: int64

----

###  Задание 6.5 ###

Какой тип клиентов (столбец `usertype`) является преобладающим — `Subscriber` или `Customer`? В качестве ответа записать долю клиентов преобладающего типа среди общего количества клиентов. Ответ округлить до сотых.

***Ответ - 0.77***

In [56]:
display(bike_df['usertype'].value_counts())
round(bike_df[bike_df['usertype'] == 'Subscriber'].shape[0] / 300_000, 2)

Subscriber    232202
Customer       67798
Name: usertype, dtype: int64

0.77

----

###  Задание 6.6 ###

Кто больше занимается велоспортом — мужчины или женщины? В ответ записать число поездок для той группы, у которой их больше.

***Ответ - 183582***

In [60]:
display(bike_df['gender'].value_counts().iloc[0])

183582

----

###  Задание 6.7 ###

Выбрать утверждения, которые соответствуют нашим данным:
1. Число уникальных стартовых и конечных стоянок, которыми воспользовались клиенты, не равны друг другу
2. В рассматриваемые дни минимальный возраст клиента составлял 10 лет (на сентябрь 2018 года)
3. Самой непопулярной стартовой стоянкой из тех, которыми воспользовались клиенты, является стоянка с названием "Eastern Pkwy & Washington Ave"
4. Наибольшее количество поездок завершается на стоянке под названием "Liberty Light Rail"


In [62]:
display(bike_df['start station id'].nunique())
display(bike_df['start station name'].nunique())
display(bike_df['end station id'].nunique())
display(bike_df['end station name'].nunique())

759

759

765

765

In [75]:
bike_df[bike_df['birth year'] > 2001 ].shape[0]
bike_df[bike_df['birth year'] > 2007 ].shape[0]

0

In [79]:
bike_df['start station name'].value_counts()

Grand Army Plaza & Central Park S    1928
Central Park S & 6 Ave               1909
Pershing Square North                1873
12 Ave & W 40 St                     1845
West St & Chambers St                1835
                                     ... 
NYCBS Depot - GOW                      15
Franklin Ave & Empire Blvd             11
Railroad Ave & Kay Ave                  9
47 Ave & Skillman Ave                   7
Eastern Pkwy & Washington Ave           6
Name: start station name, Length: 759, dtype: int64

In [80]:
bike_df['end station name'].mode()

0    West St & Chambers St
Name: end station name, dtype: object

----

###  Задание 6.8 ###

В первую очередь надо удалить лишнюю информацию из данных.

В данных присутствуют столбцы, которые дублируют информацию друг о друге: это столбцы с идентификатором и названием стартовой и конечной стоянки. Удалить признаки идентификаторов стоянок. Сколько столбцов осталось?

***Ответ - 12***

In [85]:
bike_df.drop(['start station id', 'end station id'], axis='columns', inplace=True)
display(bike_df.info())
display(bike_df.shape[1])

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300000 entries, 0 to 299999
Data columns (total 12 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   starttime                300000 non-null  object 
 1   stoptime                 300000 non-null  object 
 2   start station name       299831 non-null  object 
 3   start station latitude   300000 non-null  float64
 4   start station longitude  300000 non-null  float64
 5   end station name         299831 non-null  object 
 6   end station latitude     300000 non-null  float64
 7   end station longitude    300000 non-null  float64
 8   bikeid                   300000 non-null  int64  
 9   usertype                 300000 non-null  object 
 10  birth year               300000 non-null  int64  
 11  gender                   300000 non-null  int64  
dtypes: float64(4), int64(3), object(5)
memory usage: 27.5+ MB


None

12

----

###  Задание 6.9 ###

Заменить признак `birth year` на более понятный признак возраста клиента `age`. Годом отсчёта возраста выбрать 2018 год. Столбец `birth year` удалить из таблицы. Сколько поездок совершено клиентами старше 60 лет?

***Ответ - 11837***

In [97]:
bike_df['age'] = 2018 - bike_df['birth year']
bike_df.drop('birth year', axis='columns', inplace=True)
display(bike_df.info())
##### display(bike_df[bike_df['age'] > 60]['age'].value_counts())
display(bike_df[bike_df['age'] > 60].shape[0])

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300000 entries, 0 to 299999
Data columns (total 12 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   starttime                300000 non-null  object 
 1   stoptime                 300000 non-null  object 
 2   start station name       299831 non-null  object 
 3   start station latitude   300000 non-null  float64
 4   start station longitude  300000 non-null  float64
 5   end station name         299831 non-null  object 
 6   end station latitude     300000 non-null  float64
 7   end station longitude    300000 non-null  float64
 8   bikeid                   300000 non-null  int64  
 9   usertype                 300000 non-null  object 
 10  gender                   300000 non-null  int64  
 11  age                      300000 non-null  int64  
dtypes: float64(4), int64(3), object(5)
memory usage: 27.5+ MB


None

11837

----

###  Задание 6.10 ###

Создать признак длительности поездки `trip duration`. Для этого вычислить интервал времени между временем окончания поездки (stoptime) и её началом (starttime). Сколько целых минут длилась поездка под индексом 3 в таблице?

***Ответ - 7***

In [110]:
display(bike_df['starttime'])
display(bike_df['stoptime'])
bike_df['trip duration'] = pd.to_datetime(bike_df['stoptime']) - pd.to_datetime(bike_df['starttime'])
display(bike_df.info())
display(bike_df['trip duration'].iloc[3])

0         2018-09-01 00:00:05.2690
1         2018-09-01 00:00:11.2810
2         2018-09-01 00:00:20.6490
3         2018-09-01 00:00:21.7460
4         2018-09-01 00:00:27.3150
                    ...           
299995    2018-09-05 19:08:27.8460
299996    2018-09-05 19:08:28.3700
299997    2018-09-05 19:08:27.5090
299998    2018-09-05 19:08:29.2300
299999    2018-09-05 19:08:29.7110
Name: starttime, Length: 300000, dtype: object

0         2018-09-01 00:27:20.6340
1         2018-09-01 00:02:23.4810
2         2018-09-01 00:55:58.5470
3         2018-09-01 00:07:38.5830
4         2018-09-01 02:21:25.3080
                    ...           
299995    2018-09-05 19:15:51.4940
299996    2018-09-05 19:20:01.5080
299997    2018-09-05 19:13:40.5060
299998    2018-09-05 20:04:29.3220
299999    2018-09-05 19:12:14.7620
Name: stoptime, Length: 300000, dtype: object

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300000 entries, 0 to 299999
Data columns (total 13 columns):
 #   Column                   Non-Null Count   Dtype          
---  ------                   --------------   -----          
 0   starttime                300000 non-null  object         
 1   stoptime                 300000 non-null  object         
 2   start station name       299831 non-null  object         
 3   start station latitude   300000 non-null  float64        
 4   start station longitude  300000 non-null  float64        
 5   end station name         299831 non-null  object         
 6   end station latitude     300000 non-null  float64        
 7   end station longitude    300000 non-null  float64        
 8   bikeid                   300000 non-null  int64          
 9   usertype                 300000 non-null  object         
 10  gender                   300000 non-null  int64          
 11  age                      300000 non-null  int64          
 12  tr

None

Timedelta('0 days 00:07:16.837000')

----

###  Задание 6.11 ###

Создать «признак-мигалку» `weekend`, который равен 1, если поездка начиналась в выходной день (суббота или воскресенье), и 0 — в противном случае. Выясните, сколько поездок начиналось в выходные.

***Ответ - 115135***

In [119]:
bike_df['weekend'] = pd.to_datetime(bike_df['starttime']).dt.dayofweek
display(bike_df.info())
display(bike_df['weekend'].value_counts())
print(bike_df[bike_df['weekend'] > 4].shape[0])

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300000 entries, 0 to 299999
Data columns (total 14 columns):
 #   Column                   Non-Null Count   Dtype          
---  ------                   --------------   -----          
 0   starttime                300000 non-null  object         
 1   stoptime                 300000 non-null  object         
 2   start station name       299831 non-null  object         
 3   start station latitude   300000 non-null  float64        
 4   start station longitude  300000 non-null  float64        
 5   end station name         299831 non-null  object         
 6   end station latitude     300000 non-null  float64        
 7   end station longitude    300000 non-null  float64        
 8   bikeid                   300000 non-null  int64          
 9   usertype                 300000 non-null  object         
 10  gender                   300000 non-null  int64          
 11  age                      300000 non-null  int64          
 12  tr

None

1    68628
2    61185
6    60812
0    55052
5    54323
Name: weekend, dtype: int64

115135


----

###  Задание 6.12 ###

Создать признак времени суток поездки `time_of_day`. Время суток определять из часа начала поездки. То есть:
- поездка совершается ночью (`night`), если её час приходится на интервал от 0 (включительно) до 6 (включительно) часов;
- поездка совершается утром (`morning`), если её час приходится на интервал от 6 (не включительно) до 12 (включительно) часов;
- поездка совершается днём (`day`), если её час приходится на интервал от 12 (не включительно) до 18 (включительно) часов;
- поездка совершается вечером (`evening`), если её час приходится на интервал от 18 (не включительно) до 23 часов (включительно).

Во сколько раз количество поездок, совершённых днём, больше, чем количество поездок, совёршенных ночью, за представленный в данных период времени? Ответ округлить до целых.

***Ответ - 9***

In [129]:
bike_df['starttime'] = pd.to_datetime(bike_df['starttime'])
bike_df['stoptime'] = pd.to_datetime(bike_df['stoptime'])

In [141]:
def set_tod(arg):
    if 0 <= arg <= 6:
        return 'night'
    elif 6 < arg <= 12:
        return 'morning'
    elif 12 < arg <= 18:
        return 'day'
    else:
        return 'evening'

bike_df['time_of_day'] = bike_df['starttime'].dt.hour.apply(set_tod)
# display(bike_df.info())
display(bike_df.time_of_day.value_counts())
print('Ответ -',
      round(bike_df[bike_df.time_of_day == 'day'].shape[0]
            / bike_df[bike_df.time_of_day == 'night'].shape[0]))

day        143012
morning     95530
evening     46373
night       15085
Name: time_of_day, dtype: int64

Ответ - 9
